File: benchmarks3d/convex_polyhedron3.rs
use na::Point3;
use rand::distributions::{Distribution, Standard};
use rand::{rngs::StdRng, SeedableRng};
use rapier3d::dynamics::{JointSet, RigidBodyBuilder, RigidBodySet};
use rapier3d::geometry::{ColliderBuilder, ColliderSet};
use rapier_testbed3d::Testbed;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let joints = JointSet::new();

    /*
     * Ground
     */
    let ground_size = 200.1;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::new_static()
        .translation(0.0, -ground_height, 0.0)
        .build();
    let handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size).build();
    colliders.insert(collider, handle, &mut bodies);

    /*
     * Create the cubes
     */
    let num = 8;
    let scale = 2.0;
    let rad = 1.0;
    let border_rad = 0.1;

    let shift = border_rad * 2.0 + scale;
    let centerx = shift * (num / 2) as f32;
    let centery = shift / 2.0;
    let centerz = shift * (num / 2) as f32;

    let mut offset = -(num as f32) * shift * 0.5;

    let mut rng = StdRng::seed_from_u64(0);
    let distribution = Standard;

    for j in 0usize..47 {
        for i in 0..num {
            for k in 0usize..num {
                let x = i as f32 * shift - centerx + offset;
                let y = j as f32 * shift + centery + 3.0;
                let z = k as f32 * shift - centerz + offset;

                let mut points = Vec::new();
                for _ in 0..10 {
                    let pt: Point3<f32> = distribution.sample(&mut rng);
                    points.push(pt * scale);
                }

                // Build the rigid body.
                let rigid_body = RigidBodyBuilder::new_dynamic().translation(x, y, z).build();
                let handle = bodies.insert(rigid_body);
                let collider = ColliderBuilder::round_convex_hull(&points, border_rad)
                    .unwrap()
                    .build();
                colliders.insert(collider, handle, &mut bodies);
            }
        }

        offset -= 0.05 * rad * (num as f32 - 1.0);
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, joints);
    testbed.look_at(Point3::new(100.0, 100.0, 100.0), Point3::origin());
}

fn main() {
    let testbed = Testbed::from_builders(0, vec![("Boxes", init_world)]);
    testbed.run()
}

File: python/rapier-testbed/rapier_testbed/examples3/stress_tests/convex_polyhedron3.py
"""Port of examples3d/stress_tests/convex_polyhedron3.rs.

The Rust example uses ``ColliderBuilder::round_convex_hull`` which is not
exposed in Python; we fall back to :meth:`Collider.convex_hull` (no
border-radius rounding) which is functionally equivalent for the stress
test.
"""
from __future__ import annotations

import numpy as np

import rapier3d as rp
from ..._registry import register

CATEGORY = "Stress Tests"
NAME = "Convex polyhedron"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 200.1
    ground_height = 0.1
    ground = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_h = bodies.insert(ground)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size), ground_h, bodies
    )

    num = 8
    scale = 2.0
    rad = 1.0
    border_rad = 0.1
    shift = border_rad * 2.0 + scale
    centerx = shift * (num // 2)
    centery = shift / 2.0
    centerz = shift * (num // 2)
    offset = -float(num) * shift * 0.5

    rng = np.random.default_rng(0)
    for j in range(47):
        for i in range(num):
            for k in range(num):
                x = i * shift - centerx + offset
                y = j * shift + centery + 3.0
                z = k * shift - centerz + offset
                pts = (rng.random((10, 3)).astype(np.float32)) * scale
                body = rp.RigidBody.dynamic().translation((x, y, z))
                h = bodies.insert(body)
                col = rp.Collider.convex_hull(pts)
                if col is None:
                    continue
                colliders.insert_with_parent(col, h, bodies)
        offset -= 0.05 * rad * (num - 1)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((100.0, 100.0, 100.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from ..._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
